File: SkiaSharpImageViewer/CSharpProj/SkiaSharpImageViewer/ImageLoader.cs
using System;
using System.Collections.Generic;
using System.Drawing;

namespace SkiaSharpImageViewer
{
    public abstract class ImageLoader
    {
        protected static Dictionary<string, Image> imageCache = new Dictionary<string, Image>(); // 캐시
        private const int MaxCacheSize = 50; // 최대 캐시 크기

        public virtual Image LoadImage(string filePath)
        {
            if (TryGetFromCache(filePath, out var cachedImage))
            {
                return cachedImage;
            }

            var image = LoadImageInternal(filePath);
            AddToCache(filePath, image);
            return image;
        }

        protected bool TryGetFromCache(string filePath, out Image cachedImage)
        {
            return imageCache.TryGetValue(filePath, out cachedImage);
        }

        protected void AddToCache(string filePath, Image image)
        {
            if (imageCache.Count >= MaxCacheSize)
            {
                // 가장 오래된 항목 제거 (FIFO)
                var oldestKey = imageCache.Keys.GetEnumerator().Current;
                imageCache.Remove(oldestKey);
            }

            imageCache[filePath] = image;
        }

        protected abstract Image LoadImageInternal(string filePath);
    }
}

File: SkiaSharpImageViewer/PascalABCdotNetProj/SkiaSharpImageViewer/ImageLoader.pas
unit ImageLoader;

uses System.Drawing;

var
  imageCache: Dictionary<string, Image>; // 전역 변수로 캐시 선언
  
type
  ImageLoader2 = abstract class
  private
    const MaxCacheSize = 50; // 최대 캐시 크기
    //class var imageCache: Dictionary<string, Image>; // 캐시
    
    // 가장 오래된 항목 제거 (FIFO)
    class procedure RemoveOldestFromCache;
    begin
      var oldestKey := imageCache.Keys.First;
      imageCache.Remove(oldestKey);
    end;
    
  public
    constructor;
    begin
      if imageCache = nil then
        imageCache := new Dictionary<string, Image>;
    end;
    
    // 캐시에서 이미지 가져오기 시도
    function TryGetFromCache(filePath: string; var cachedImage: Image): boolean;
    begin
      Result := imageCache.TryGetValue(filePath, cachedImage);
    end;
    
    // 캐시에 이미지 추가
    procedure AddToCache(filePath: string; image: Image);
    begin
      if imageCache.Count >= MaxCacheSize then
        RemoveOldestFromCache;
      
      imageCache[filePath] := image;
    end;
    
    // 이미지 로드 메서드
    function LoadImage(filePath: string): Image;
    var
      cachedImage: Image;
    begin
      if TryGetFromCache(filePath, cachedImage) then
      begin
        Result := cachedImage;
        exit;
      end;
      
      var loadedImage := LoadImageInternal(filePath);
      AddToCache(filePath, loadedImage);
      Result := loadedImage;
    end;
    
    // 추상 메서드: 하위 클래스에서 구현 필요
    protected function LoadImageInternal(filePath: string): Image; abstract;
  end;
  
  begin
    
  
  end.